llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp

```diff
@@ -305,6 +306,48 @@ static Value *expandNormalizeIntrinsic(CallInst *Orig) {
   return Builder.CreateFMul(X, MultiplicandVec);
 }
 
+static Value *expandAtan2Intrinsic(CallInst *Orig) {
+  Value *Y = Orig->getOperand(0);
+  Value *X = Orig->getOperand(1);
+  Type *Ty = X->getType();
+  IRBuilder<> Builder(Orig);
+
+  Value *Tan = Builder.CreateFDiv(Y, X);
+
+  Value *Atan =
+      Builder.CreateIntrinsic(Ty, Intrinsic::atan, {Tan}, nullptr, "Elt.Atan");
+
+  Constant *Pi = ConstantFP::get(Ty, llvm::numbers::pi);
+  Constant *HalfPi = ConstantFP::get(Ty, llvm::numbers::pi / 2);
+  Constant *NegHalfPi = ConstantFP::get(Ty, -llvm::numbers::pi / 2);
+  Constant *Zero = ConstantFP::get(Ty, 0);
+
+  Value *AtanAddPi = Builder.CreateFAdd(Atan, Pi);
+  Value *AtanSubPi = Builder.CreateFSub(Atan, Pi);
+
+  Value *Result = Atan;
+
+  Value *XLt0 = Builder.CreateFCmpOLT(X, Zero);
+  Value *XEq0 = Builder.CreateFCmpOEQ(X, Zero);
+
+  Value *YGe0 = Builder.CreateFCmpOGE(Y, Zero);
+  Value *YLt0 = Builder.CreateFCmpOLT(Y, Zero);
+
+  Value *XLt0AndYGe0 = Builder.CreateAnd(XLt0, YGe0);
+  Result = Builder.CreateSelect(XLt0AndYGe0, AtanAddPi, Result);
+
+  Value *XLt0AndYLt0 = Builder.CreateAnd(XLt0, YLt0);
+  Result = Builder.CreateSelect(XLt0AndYLt0, AtanSubPi, Result);
+
+  Value *XEq0AndYLt0 = Builder.CreateAnd(XEq0, YLt0);
+  Result = Builder.CreateSelect(XEq0AndYLt0, NegHalfPi, Result);
+
+  Value *XEq0AndYGe0 = Builder.CreateAnd(XEq0, YGe0);
+  Result = Builder.CreateSelect(XEq0AndYGe0, HalfPi, Result);
+
+  return Result;
```

Comment:
We typically do something like this at the end.
```c++
  newCall->setTailCall(Orig->isTailCall());
  newCall->setAttributes(Orig->getAttributes());
```
Is thee a reason we aren't doing that here?

Reply:
Looking for that, I see three examples (for log, pow, exp), but this doesn't not seem to be done *most* of the time.  Where it is done, it appears to be only applied to the last intrinsic call, not all of the expanded calls.

Do you know whether these are necessary for our intrinsic expansions?  The thing I would expect to be important would be to preserve fast math flags and FPMathTag, which these don't do either.

For TailCall, it doesn't seem like anything we could use or want for DXIL ops anyway.
For other attributes, what attributes would need to be preserved from the original intrinsic call, and how should these apply to all the instructions and DXIL op calls generated?

Reply:
Correct you would not apply it to the enitre expansion just to atan2. Second those examples are all examples where we go from an llvm intrinsic to another llvm intrinsic which atan2 to atan is doing, hence why I ask the question. the idea would be that if atan2 had some attributes that atan did not have then we would copy them over.

Reply:
I still don't know what of value we could actually be preserving, and whether everything else is missing the preservation of fast math flags, etc...
There are various other intrinsic expansions that don't do this when creating a new dx intrinsic (like abs, dot, length, normalize, clamp).  Wouldn't this also apply to all intrinsic calls generated in an expansion (like log2 from pow expansion)?

Reply:
I did add this, so we can decide what we want to do and apply that consistently to the whole set later.  Let me know if I should do anything more here.

Reply:
create a cleanup task and we will track it in the backlog.

Reply:
I filed this issue for tracking:
HLSL intrinsic expansion cleanup - preserve/apply attributes/tail call/FMF consistently #109717